### PR TITLE
feat: implement reproducible build and get cached image

### LINKED ISSUE #186: PoC: Build a tool to validate file hashes against layer caches
This issue tracks the implementation of a PoC to validate the path forward for #128.

To support #185, we must be able to figure out if a cached layer image is valid, given the state  of files relevant to building the container (think contents of `Dockerfile`, `devcontainer.json`, and hashes of any files pulled in via `COPY`-directive).

In this PoC, we will implement this logic (files -> layer) as a subcommand of envbulider. This will require parsing and understanding of `Dockerfile` to understand if a change modifies the outcome, and which layers remain intact.

### PR DESCRIPTION
**Pull request by mafredri**
This PR is a PoC, it builds upon #197 and https://github.com/coder/kaniko/pull/12 and shows how build cache + pushed image can be utilized.

Example get-cached-image:

```
❯ registry_ip=$(docker inspect envbuilder-registry | jq -r '.[].NetworkSettings.IPAddress'); docker run -it --rm -v /tmp/envbuilder:/workspaces -e ENVBUILDER_CACHE_REPO=${registry_ip}:5000/local/cache -e ENVBUILDER_GIT_URL=https://github.com/coder/envbuilder-starter-devcontainer -e ENVBUILDER_INIT_SCRIPT=exit -e ENVBUILDER_GET_CACHED_IMAGE=true envbuilder:latest
envbuilder - Build development environments from repositories in a container
#1: 📦 Cloning https://github.com/coder/envbuilder-starter-devcontainer to /workspaces/envbuilder-starter-devcontainer...
...
2: RUN apt-get update
2: unable to build fake image: error fake building stage: uncached command *commands.RunMarkerCommand is not supported in fake build
```

Now, do build:

```
❯ registry_ip=$(docker inspect envbuilder-registry | jq -r '.[].NetworkSettings.IPAddress'); docker run -it --rm -v /tmp/envbuilder:/workspaces -e ENVBUILDER_CACHE_REPO=${registry_ip}:5000/local/cache -e ENVBUILDER_GIT_URL=https://github.com/coder/envbuilder-starter-devcontainer -e ENVBUILDER_INIT_SCRIPT=exit -e ENVBUILDER_PUSH_IMAGE=true envbuilder:latest
envbuilder - Build development environments from repositories in a container
#1: 📦 Cloning https://github.com/coder/envbuilder-starter-devcontainer to /workspaces/envbuilder-starter-devcontainer...
...
2: Running: [/bin/sh -c echo "PS1='🐳 \[\033[1;36m\]\h \[\033[1;34m\]\W\[\033[0;35m\] \[\033[1;36m\]# \[\033[0m\]'" > ~/.bashrc]
2: Taking snapshot of files...
2: Pushing layer 172.17.0.3:5000/local/cache:445691a7ed1a9ce9faac28095b3f3b77a0043c7604b099c299bd4abed5089f57 to cache now
2: Pushing image to 172.17.0.3:5000/local/cache:445691a7ed1a9ce9faac28095b3f3b77a0043c7604b099c299bd4abed5089f57
2: Pushed 172.17.0.3:5000/local/cache@sha256:440a1a035eb3ba7597aa7b3f1b5e08c67c3c5afed33bea057dd18640420ca115
#2: 🏗️ Built image! [16.509477016s]
#3: 🏗️ Pushing image...
3: Pushing image to 172.17.0.3:5000/local/cache
3: Pushed 172.17.0.3:5000/local/cache@sha256:9c102cd02585011898abd11ea367ba5368415b9cd9ecebbb07dcbb57f6e4cd63
#3: 🏗️ Pushed image! [1.003674439s]
=== Running the init command /bin/sh [-c exit] as the "coder" user...
```

Try get-cached-image again:

```
❯ registry_ip=$(docker inspect envbuilder-registry | jq -r '.[].NetworkSettings.IPAddress'); docker run -it --rm -v /tmp/envbuilder:/workspaces -e ENVBUILDER_CACHE_REPO=${registry_ip}:5000/local/cache -e ENVBUILDER_GIT_URL=https://github.com/coder/envbuilder-starter-devcontainer -e ENVBUILDER_INIT_SCRIPT=exit -e ENVBUILDER_GET_CACHED_IMAGE=true envbuilder:latest
envbuilder - Build development environments from repositories in a container
#1: 📦 Cloning https://github.com/coder/envbuilder-starter-devcontainer to /workspaces/envbuilder-starter-devcontainer...
...
#2: 🏗️ Building fake image...
2: Retrieving image manifest ubuntu
2: Retrieving image ubuntu from registry index.docker.io
2: Built cross stage deps: map[]
2: Retrieving image manifest ubuntu
2: Returning cached image manifest
2: Executing 0 build triggers
2: Checking for cached layer 172.17.0.3:5000/local/cache:bd0def545b590c39144a57152c120a5baf8e3cad7c70298bbacfa5116ef1ca3c...
2: Using caching version of cmd: RUN apt-get update
2: Checking for cached layer 172.17.0.3:5000/local/cache:591f8c35a55cc60f7bc49fa3f2c99360616a9b6b35908b8aaa7392e7820b0c36...
2: Using caching version of cmd: RUN apt-get install vim sudo curl -y
2: Checking for cached layer 172.17.0.3:5000/local/cache:1cb1d8b00537df6f0638913717931b76e7b80b8c150448ce4096f4eef042b355...
2: Using caching version of cmd: RUN useradd -m -s /bin/bash -G sudo coder
2: Cmd: USER
2: Checking for cached layer 172.17.0.3:5000/local/cache:445691a7ed1a9ce9faac28095b3f3b77a0043c7604b099c299bd4abed5089f57...
2: Using caching version of cmd: RUN echo "PS1='🐳 \[\033[1;36m\]\h \[\033[1;34m\]\W\[\033[0;35m\] \[\033[1;36m\]# \[\033[0m\]'" > ~/.bashrc
2: RUN apt-get update
2: Found cached layer, faking extraction to filesystem
2: RUN apt-get install vim sudo curl -y
2: Found cached layer, faking extraction to filesystem
2: RUN useradd -m -s /bin/bash -G sudo coder
2: Found cached layer, faking extraction to filesystem
2: USER coder
2: Cmd: USER
2: No files changed in this command, skipping snapshotting.
2: RUN echo "PS1='🐳 \[\033[1;36m\]\h \[\033[1;34m\]\W\[\033[0;35m\] \[\033[1;36m\]# \[\033[0m\]'" > ~/.bashrc
2: Found cached layer, faking extraction to filesystem
2: Mapping stage idx 0 to digest sha256:3ff463dd27a13ebcdfc74fc790e7a8da0b5373110b4edd1dc1426c4b4ed80e27
2: Mapping digest sha256:3ff463dd27a13ebcdfc74fc790e7a8da0b5373110b4edd1dc1426c4b4ed80e27 to cachekey 445691a7ed1a9ce9faac28095b3f3b77a0043c7604b099c299bd4abed5089f57
#2: 🏗️ Built fake image! [1.298919769s]
172.17.0.3:5000/local/cache@sha256:9c102cd02585011898abd11ea367ba5368415b9cd9ecebbb07dcbb57f6e4cd63
```

And finally:

```
❯ docker run -it --rm --entrypoint /usr/bin/bash localhost:5000/local/cache@sha256:9c102cd02585011898abd11ea367ba5368415b9cd9ecebbb07dcbb57f6e4cd63
Unable to find image 'localhost:5000/local/cache@sha256:9c102cd02585011898abd11ea367ba5368415b9cd9ecebbb07dcbb57f6e4cd63' locally
localhost:5000/local/cache@sha256:9c102cd02585011898abd11ea367ba5368415b9cd9ecebbb07dcbb57f6e4cd63: Pulling from local/cache
49b384cc7b4a: Already exists
4e5c7f007bcd: Pull complete
b3cda1764760: Pull complete
62220e490978: Pull complete
07141b433862: Pull complete
Digest: sha256:9c102cd02585011898abd11ea367ba5368415b9cd9ecebbb07dcbb57f6e4cd63
Status: Downloaded newer image for localhost:5000/local/cache@sha256:9c102cd02585011898abd11ea367ba5368415b9cd9ecebbb07dcbb57f6e4cd63
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

🐳 e54cf491d1a7 / #
```

Closes #186.

### CODE REVIEW

File: options.go [hunk 1]
```diff
@@ -395,6 +402,20 @@ func (o *Options) CLI() serpent.OptionSet {
 			Description: "Coder agent subsystems to report when forwarding logs. " +
 				"The envbuilder subsystem is always included.",
 		},
+		{
+			Flag:  "push-image",
+			Env:   WithEnvPrefix("PUSH_IMAGE"),
+			Value: serpent.BoolOf(&o.PushImage),
+			Description: "Flag to determine if the image should be pushed to " +
+				"the container registry. This option implies reproducible builds.",
```
Comment (johnstcn) [inline options.go:410]: Nit: simplify wording

```suggestion
			Description: "Push the built image to a remote registry. " +
				"This option forces a reproducible build.",
```

File: options.go [hunk 2]
```diff
@@ -395,6 +402,20 @@ func (o *Options) CLI() serpent.OptionSet {
...
+		},
+		{
+			Flag:  "get-cached-image",
+			Env:   WithEnvPrefix("GET_CACHED_IMAGE"),
+			Value: serpent.BoolOf(&o.GetCachedImage),
+			Description: "Flag to determine if the cached image is available, " +
+				"and if it is, to return it.",
```
Comment (johnstcn) [inline options.go:417]: Nit: simplify wording
```suggestion
			Description: "Print the digest of the cached image, if available.",
```

File: envbuilder.go [hunk 3]
```diff
@@ -474,20 +474,24 @@ func Run(ctx context.Context, options Options) error {
...
+		opts := &config.KanikoOptions{
 			// Boilerplate!
 			CustomPlatform:    platforms.Format(platforms.Normalize(platforms.DefaultSpec())),
 			SnapshotMode:      "redo",
 			RunV2:             true,
 			RunStdout:         stdoutWriter,
 			RunStderr:         stderrWriter,
-			Destinations:      []string{"local"},
+			Destinations:      destinations,
+			NoPush:            !options.PushImage || len(destinations) == 0,
```
Comment (johnstcn) [inline envbuilder.go:494]: I'm assuming the reason for the logical disjuction here is that Kaniko will push all intermediate cache layers to `destinations`, while specifying `options.PushImage` signifies to push only the final result?

File: envbuilder.go [hunk 4]
```diff
@@ -518,11 +522,42 @@ func Run(ctx context.Context, options Options) error {
...
+
+			// For cached image utilization, produce reproducible builds.
+			Reproducible: options.PushImage,
+		}
+
+		if options.GetCachedImage {
+			endStage := startStage("🏗️ Building fake image...")
+			image, err := executor.DoFakeBuild(opts)
+			if err != nil {
+				logrus.Infof("unable to build fake image: %s", err)
```
Comment (johnstcn) [inline envbuilder.go:534]: I think the messaging here is slightly misleading. It's not an "error" as such, but we asked envbuilder to check if all the image layers are available in the cache and they're not.

File: envbuilder.go (hunk 3 above)
Comment (mafredri) [reply envbuilder.go:494]: Yep 👍🏻

File: envbuilder.go [hunk 5]
```diff
@@ -518,11 +522,42 @@ func Run(ctx context.Context, options Options) error {
 				RegistryMirrors: registryMirror,
 			},
 			SrcContext: buildParams.BuildContext,
-		})
+
+			// For cached image utilization, produce reproducible builds.
+			Reproducible: options.PushImage,
```
Comment (BrunoQuaresma) [inline envbuilder.go:527]: Where is this getting used?

Comment (johnstcn) [reply envbuilder.go:527]: See https://github.com/coder/kaniko/pull/12

File: envbuilder.go [hunk 6]
```diff
@@ -474,20 +474,24 @@ func Run(ctx context.Context, options Options) error {
...
 		// At this point we have all the context, we can now build!
 		registryMirror := []string{}
 		if val, ok := os.LookupEnv("KANIKO_REGISTRY_MIRROR"); ok {
 			registryMirror = strings.Split(val, ";")
 		}
-		image, err := executor.DoBuild(&config.KanikoOptions{
+		var destinations []string
+		if options.CacheRepo != "" {
+			destinations = append(destinations, options.CacheRepo)
+		}
```
Comment (johnstcn) [inline envbuilder.go:485]: specifying `PUSH_IMAGE` without `CACHE_REPO` should fail-fast, as this is probably not what folks want.

**Comment by kylecarbs** [comment]
I'd love to try this in dogfood asap with our devcontainer!

**Comment by kylecarbs** [comment]
And since we added support for sysbox, this might _just work_!

**Comment by kylecarbs** [comment]
I think we try using the Docker Registry to see what the performance looks like with a remote!

File: envbuilder.go [hunk 7]
```diff
@@ -99,6 +99,9 @@ func Run(ctx context.Context, options Options) error {
 	if options.InitCommand == "" {
 		options.InitCommand = "/bin/sh"
 	}
+	if options.CacheRepo == "" && options.PushImage {
+		return fmt.Errorf("--cache-repo must be set when using --push-image!")
+	}
```
Comment (johnstcn) [inline envbuilder.go:104]: review: I don't see any good reason to not do this. It prevents potential head-scratching if someone sets `ENVBUILDER_PUSH_IMAGE` without specifying `ENVBUILDER_CACHE_REPO`.

**Comment by johnstcn** [comment]
@dannykopping @BrunoQuaresma @mtojek I have updated this branch with integration tests.

It's worth noting that `--get-cached-image` currently does not work with multi-stage builds. I have filed a [separate issue](https://github.com/coder/envbuilder/issues/230) for that.

**Review by mtojek — APPROVED** [review]
We agreed on `DoCacheProbe` in kaniko's PR, so this round is fairly quick. Approved 👍

**Review by dannykopping — APPROVED** [review]
Only nits, great work!

File: integration/integration_test.go [hunk 8]
```diff
@@ -987,6 +989,252 @@ COPY %s .`, testImageAlpine, inclFile)
...
+		_, err = remote.Image(ref)
+		require.ErrorContains(t, err, "NAME_UNKNOWN", "expected image to not be present before build + push")
+
+		// When: we run envbuilder with GET_CACHED_IMAGE
+		_, err = runEnvbuilder(t, options{env: []string{
+			envbuilderEnv("GIT_URL", srv.URL),
+			envbuilderEnv("CACHE_REPO", testRepo),
+			envbuilderEnv("GET_CACHED_IMAGE", "1"),
+		}})
+		require.ErrorContains(t, err, "not supported in fake build")
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply integration/integration_test.go:1024]: Good catch, will update. 👍

[the remark this answers is not in the document]
Comment (johnstcn) [reply integration/integration_test.go:1024]: https://github.com/coder/kaniko/pull/15

**Pull request merged**